### ISSUE TITLE
[jvm-packages] Spark Barrier execution mode [RFC]

### ISSUE THREAD

**Issue by trams** [dmlc/xgboost#4793]
## Purpose of the document
I want to know what community thinks about using Spark Barrier execution mode before submitting any pull requests. I understand that I may be missing some information which could block or complicate Barrier mode integration

## Barrier execution mode
In short Barrier execution mode permits to call mapPartition on RDD and ensure that
1. all tasks will launched at once. If there is not enough task slots the exception will be produced
2. tasks either all succeed or fail. Upon task failure Spark will kill all tasks and restart whole "stage" fully
Also Spark won't make any assumption that tasks don't talk to each other. Actually it is the opposite. Spark provides [BarrierTaskContext](https://spark.apache.org/docs/latest/api/java/org/apache/spark/BarrierTaskContext.html) which facilitate all tasks discovery (and have other goodies)

## xgboost-spark
xgboost-spark already integrates xgboost using `mapPartition` so the change to start using Barrier mode is trivial. It permits us to remove our listener which closes SparkContext in case of a failure.

It also potentially permits us to restart the training from the last checkpoint in case of a failure.

We may also consider using `BarrierTaskContext` information to initialize RABIT which may result in our code simplification (I need to dig deeper here to assess this)

## Future of Spark Barrier Execution Mode
It is true that in Spark 2.4 this API has been marked `Experimental`
But from what I know it is here to stay. Why do I think so?
1. Databricks invests in Horovod which is integrated with Spark using Barrier mode
2. That's what Databricks developers (>75% of contributions to Spark) told me when I asked whether this API will be removed or not.

What do you think?

**Comment by trams** [comment]
Spark Barrier Execution Mode was mentioned https://github.com/dmlc/xgboost/issues/3996 as one of the reasons to upgrade to Spark 2.4

**Comment by trams** [comment]
In case you miss, @CodingCat, @hcho3 
I want your opinion about using Spark Barrier execution mode. If there are some interest and no opposition I will allocate some time on this and I will submit a pull request

What do you think?

**Comment by hcho3** [comment]
I'll defer to @CodingCat and @chenqin

**Comment by chenqin** [comment]
My impression Barrier Execution Mode is still in early stage, the way it were implemented were "hold and count" strategy. Can we get more clarify what happens if user turn on dynamic allocation along side with barrier execution mode?  We had seen issue with spark listener function not as expected which made us cautious of unproven features.

If not urgent, we might get this implemented in Spark 3 which would expect quite a bit more features complete e.g s per stage resource allocation https://issues.apache.org/jira/browse/SPARK-27495. This can be very helpful to work side by side with barrier execution. Support binding GPUs or giving larger native memory only in training stage.

Regarding to Horovod, ASAIK, it's pure Pyspark play. I knew they were having some challenges to integrate with Spark last year. Looks like Spark 3 seems try to address some of concerns. Some of fundamental challenges I learned by involving support horovod on spark in side of company weren't listed in spark jiras.

Reference [SPIP: Support Barrier Execution Mode in Apache Spark]https://issues.apache.org/jira/browse/SPARK-24374
[project hydrogen]https://issues.apache.org/jira/browse/SPARK-26175?jql=labels%20%3D%20Hydrogen
